### PR TITLE
[SPARK-25050][SQL] Avro: writing complex unions

### PR DESCRIPTION
**Pull request by steven-aerts**
### What changes were proposed in this pull request?

Spark was able to read complex unions already but not write them. 
Now it is possible to also write them.  If you have a schema with a complex union the following code is now working:

```scala
spark
  .read.format("avro").option("avroSchema", avroSchema).load(path)
  .write.format("avro").option("avroSchema", avroSchema).save("/tmp/b")
```
While before this patch it would throw `Unsupported Avro UNION type` when writing.

Add the capability to write complex unions, next to reading them.
Complex unions map to struct types where field names are member0, member1, etc.
This is consistent with the behavior in SchemaConverters for reading them
and when converting between Avro and Parquet.

### Why are the changes needed?
Fixes SPARK-25050, lines up read and write compatibility.

### Does this PR introduce _any_ user-facing change?
The behaviour improved of course, this is as far as I could see not impacting any customer facing API's or documentation.

### How was this patch tested?
- Added extra unit tests.
- Updated existing unit tests for improved behaviour.
- Validated manually with an internal corpus of avro files if they now could be read and written without problems.  Which was not before this patch.

### CODE REVIEW

**Comment by thejdeep** [comment]
@steven-aerts Thanks for working on this feature.

+1 to this PR. The lack of complex union type write support causes us problems too. Right now, since the standard Dataframe/Dataset APIs do not support writing out unions with multiple subtypes, we have been deferring to changing the underlying schema which maybe cumbersome in some cases or having to use the [saveAsNewAPIHadoopFile](https://spark.apache.org/docs/3.0.0/api/scala/org/apache/spark/rdd/PairRDDFunctions.html#saveAsNewAPIHadoopFile(path:String,keyClass:Class%5B_%5D,valueClass:Class%5B_%5D,outputFormatClass:Class%5B_%3C:org.apache.hadoop.mapreduce.OutputFormat%5B_,_%5D%5D,conf:org.apache.hadoop.conf.Configuration):Unit) RDD API which skips the Catalyst path.

**Comment by mridulm** [comment]
+CC @dongjoon-hyun, @HyukjinKwon who might be able to review this better than me.

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (xkrogen) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:230]: This does imply some loss of information if you were to do a round-trip from Avro to SQL to Avro, since all records written out would have double/long values even if the input was a float/int.

The representation chosen on the read path is inherently lossy, since we discard the information about which union branch the datum originated from, so I don't think there's anything we can do here to avoid this behavior. Just wanted to point it out.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (xkrogen) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:1222]: Maybe use `SchemaBuilder` here? It was hard for me to read the unionSchema examples with so many quotes
```suggestion
    val recordS = SchemaBuilder.record("r").fields().requiredString("f1").endRecord()
    val intS = Schema.create(Schema.Type.INT)
    val nullS = Schema.create(Schema.Type.NULL)
    for ((unionTypes, compatible) <- Seq(
      (Seq(nullS, intS, recordS), true),
      (Seq(intS, nullS, recordS), true),
      (Seq(intS, recordS, nullS), true),
      (Seq(intS, recordS), true),
      (Seq(nullS, recordS, intS), false),
      (Seq(nullS, recordS), false),
      (Seq(nullS, SchemaBuilder.record("r").fields().requiredString("f2").endRecord()), false)
    )) {
      val avroSchema = SchemaBuilder.record("test_schema").fields()
        .name("union").`type`(Schema.createUnion(unionTypes: _*)).noDefault()
        .endRecord().toString()
```

**Review by xkrogen — CHANGES_REQUESTED** [review]
Great to see this! Overall approach looks great, though I have some minor implementation comments throughout. No blockers except for making sure we properly throw an exception when we encounter a union with no non-null types.

**Comment by xkrogen** [comment]
@steven-aerts just want to check if you are still working on this?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (steven-aerts) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:230]: I think is not lost, as it as the schema has to be provided when writing avro.  If the avro schema defines a union it will be written as a union, and this info is not really lost.

**Review by steven-aerts — COMMENTED** [review]
@xkrogen thanks for your wonderful review.  I am still on vacation, but back in full force next week to take all your nice comments in.

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (xkrogen) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:230]: Yes, I agree. No need to fix here.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (xkrogen) [reply connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala]: Ah yes I see. Makes sense, thanks.

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (xkrogen) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:318]: We should respect `AvroOptions.positionalFieldMatching` here and only do this check if `positionalFieldMatching = false`, WDYT?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (xkrogen) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: Yeah that's fair. Slightly less precise, but I agree it will probably be more in-line with user expectations.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala
Comment (xkrogen) [reply connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala]: Yeah I agree the test as-is won't work. I was thinking more like adding another test for the more complex case. My point was more that the int-long/float-double case is only ... "partially complex"? And this test claims to be testing the complex case, which (to me at least) would indicate that we get an exploded record.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala
Comment (xkrogen) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroSuite.scala:349]: Can we confirm the value of `df2.selectExpr("field5")` as well (here and above)?

**Review by xkrogen — COMMENTED** [review]
Looking really good to me! Left a few comments but they are minor.

@gengliangwang would you be able to help review as well? This is a very commonly asked-for feature from our users, would love to see it go in to 3.4.0 :)

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala
Comment (xkrogen) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala:243]: Why the explicit reference to `Tuple2`? You should be able to just do `(Some(1), None)` right?

Comment (steven-aerts) [reply connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroFunctionsSuite.scala]: Done

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (robreeves) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:371]: This logic has a couple of issues determining nullability for a union. Checking if the union contains a null is needed to handle these cases.

1. Case `["null", "int", "long"]` will hit line 374 and return `nullable=false` when it should return true.
2. Case `["int"]` will hit line 373 and return `nullable=true` when it should return false.

Maybe something like this?

```scala
val containsNull = avroType.getTypes.asScala.exists(_.getType == Schema.Type.NULL)

nonNullUnionBranches(avroType) match {
    case Seq() => (true, Schema.create(Type.NULL))
    case Seq(singleType) => (containsNull, singleType)
    case _ => (containsNull, avroType)
}
```

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (robreeves) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:374]: `avroTypes` contains the null type in the union. Should this return the union types with the null filtered out? That is what it previously did for nullable unions.

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (robreeves) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:350]: This comment still says that only a `["null", "type"]` union is supported. Can you update it so it is clear all unions are supported now?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (steven-aerts) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:374]: I think this is exactly what we are doing here, by using the output of `nonNullUnionBranches`.
We make however an exception of `["null"]`.

Comment (steven-aerts) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala]: So we only make an exception for _optional types_.  For _complex unions_, we however need the null further down.

**Comment by steven-aerts** [comment]
Is there still something I can/have to do to get this patch submitted?

**Comment by xkrogen** [comment]
Unfortunately none of I / @thejdeep / @robreeves are committers, just interested parties, so we can't merge. We need a review from a committer as well.

@gengliangwang , @cloud-fan , @dongjoon-hyun , can any of you take a look at this?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala
Comment (gengliangwang) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:222]: This seems different from https://github.com/apache/spark/blob/master/connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala#L296
If only one non-null field in the union, we will create a converter on the non-null field instead of a complex one.

Comment (steven-aerts) [reply connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroSerializer.scala:222]: @gengliangwang this branch will only be hit if the resolved avro type is determined to be of the type union.

For optional fields, so a union of a type with `null`.  The [resolution ](https://github.com/apache/spark/pull/36506/files/1d950f55e85f11bfb85d9d57324d7ef7916b4a54#diff-e84c7dfce11df94178029c05a8427668c9dab4bb809ed6e362a39ff043a59f89R369) already deterimined that it was not a union type and will hit the branches another branch in this case.  It is also why `newConverter` is almost always called from a type generated by [`resolveNullableType`](https://github.com/apache/spark/pull/36506/files/1d950f55e85f11bfb85d9d57324d7ef7916b4a54#diff-e84c7dfce11df94178029c05a8427668c9dab4bb809ed6e362a39ff043a59f89R358).

Let me know if this is not clear.
Btw, I missed you question somehow, sorry for the late reply.

**Comment by xkrogen** [comment]
@gengliangwang any comments on the latest diff, after @steven-aerts answered your last question? Seems that this PR is in a very healthy state, I would love to see it merged.

**Comment by xkrogen** [comment]
Ping @cloud-fan @gengliangwang @dongjoon-hyun , are any of you available to help review?

**Comment by gengliangwang** [comment]
@steven-aerts @xkrogen Sorry for the late reply. I will take another look later today.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
cc @bozhang2820 since you made https://github.com/apache/spark/commit/551b504cfe38d1ab583e617c37e49659edd65c2e

**Comment by gengliangwang** [comment]
Merging to master/3.4. cc Spark 3.4.0 release manager @xinrong-meng

**Comment by dongjoon-hyun** [comment]
Thank you all! I'm also supporting @gengliangwang 's backporting decision.

**Pull request closed without merging**
